oximeter/collector/src/agent.rs

```diff
@@ -131,10 +144,128 @@ async fn perform_collection(
-                .datum
-                .increment()
+            Err(self_stats::FailureReason::Unreachable)
+        }
+    }
+}
+
+// The type of one collection task run to completion.
+//
+// An `Err(_)` means we failed to collect, and contains the reason so that we
+// can bump the self-stat counter accordingly.
+type CollectionResult = Result<ProducerResults, self_stats::FailureReason>;
+
+// The type of one response message sent from the collection task.
+type CollectionResponse = (Option<CollectionToken>, CollectionResult);
+
+/// Task that actually performs collections from the producer.
+async fn inner_collection_loop(
+    log: Logger,
+    mut producer_info_rx: watch::Receiver<ProducerEndpoint>,
+    mut forced_collection_rx: mpsc::Receiver<CollectionToken>,
+    mut timer_collection_rx: mpsc::Receiver<()>,
+    result_tx: mpsc::Sender<CollectionResponse>,
+) {
+    let client = reqwest::Client::builder()
+        .timeout(COLLECTION_TIMEOUT)
+        .build()
+        // Safety: `build()` only fails if TLS couldn't be initialized or the
+        // system DNS configuration could not be loaded.
+        .unwrap();
+    loop {
+        // Wait for notification that we have a collection to perform, from
+        // either the forced- or timer-collection queue.
+        trace!(log, "top of inner collection loop, waiting for next request",);
+        let maybe_token = tokio::select! {
+            maybe_request = forced_collection_rx.recv() => {
+                let Some(request) = maybe_request else {
+                    debug!(
+                        log,
+                        "forced collection request queue closed, exiting"
+                    );
+                    return;
+                };
+                Some(request)
+            }
+            maybe_request = timer_collection_rx.recv() => {
+                if maybe_request.is_none() {
+                    debug!(
+                        log,
+                        "timer collection request queue closed, exiting"
+                    );
+                    return;
+                };
+                None
+            }
+        };
+
+        // Make a future to represent the actual collection.
+        let mut collection_fut = Box::pin(perform_collection(
+            log.clone(),
+            client.clone(),
+            producer_info_rx.borrow_and_update().clone(),
+        ));
+
+        // Wait for that collection to complete or fail, or for an update to the
+        // producer's information. In the latter case, recreate the future for
+        // the collection itself with the new producer information.
+        let collection_result = 'collection: loop {
+            tokio::select! {
+                biased;
+
+                maybe_update = producer_info_rx.changed() => {
+                    match maybe_update {
+                        Ok(_) => {
+                            let update = producer_info_rx.borrow().clone();
```

Comment:
I think this should be

```suggestion
                            let update = producer_info_rx.borrow_and_update().clone();
```

(and actually, I think nearly _all_ watch channel receivers should use `borrow_and_update` instead of `borrow`; I wish they were flipped and `borrow` did the right thing and `borrow_without_update` was the longer-to-write one).

`borrow_and_update` is needed to correctly handle this ordering:

1. Producer info changes
2. We wake up from `producer_info_rx.changed()` on line 216
3. Producer info changes again
4. We `borrow()` here; we'll see the info from the second change, but borrow doesn't mark the value as seen in the channel. When we `continue 'collection`, `producer_info_rx.changed()` immediately fires again because there's been a change since the last `.changed()` call

Reply:
Done in 69dfa7a9d, thanks for the careful look.